From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tobias Werth <werth@cs.fau.de>
Date: Sat, 2 May 2020 16:38:06 +0200
Subject: cherry-pick: Refactor balloon index action.

No functional changes intended.

This is in preparation of replacing balloons.in, see issue 425.

(cherry picked from commit a7c694c93e374096f770322ffa6854674e6b9fec)
---
 .../src/Controller/Jury/BalloonController.php | 121 ++----------------
 webapp/src/Service/BalloonService.php         | 110 ++++++++++++++++
 2 files changed, 120 insertions(+), 111 deletions(-)

diff --git a/webapp/src/Controller/Jury/BalloonController.php b/webapp/src/Controller/Jury/BalloonController.php
index fe55509ee..99be9cfe0 100644
--- a/webapp/src/Controller/Jury/BalloonController.php
+++ b/webapp/src/Controller/Jury/BalloonController.php
@@ -5,6 +5,7 @@ namespace App\Controller\Jury;
 use App\Entity\Balloon;
 use App\Entity\ScoreCache;
 use App\Entity\TeamAffiliation;
+use App\Service\BalloonService;
 use App\Service\ConfigurationService;
 use App\Service\DOMJudgeService;
 use App\Service\EventLogService;
@@ -67,134 +68,32 @@ class BalloonController extends AbstractController
     /**
      * @Route("", name="jury_balloons")
      */
-    public function indexAction(Request $request, Packages $assetPackage, KernelInterface $kernel)
+    public function indexAction(Request $request, BalloonService $balloonService)
     {
-        $showPostFreeze = (bool)$this->config->get('show_balloons_postfreeze');
-
-        $em = $this->em;
 
         $contest = $this->dj->getCurrentContest();
         if(is_null($contest)) {
             return $this->render('jury/balloons.html.twig');
         }
 
-        $contestIsFrozen = isset($contest->getState()['frozen']);
-        if(!$showPostFreeze) {
-            $freezetime = $contest->getFreezeTime();
-        }
-
-        // Build a list of teams and the problems they solved first
-        $firstSolved = $em->getRepository(ScoreCache::class)->findBy(['is_first_to_solve'=>1]);
-        $firstSolvers = [];
-        foreach($firstSolved as $scoreCache) {
-            $firstSolvers[$scoreCache->getTeam()->getTeamId()][] = $scoreCache->getProblem()->getProbid();
-        }
-
-        $query = $em->createQueryBuilder()
-            ->select('b', 's.submittime', 'p.probid',
-                't.teamid', 't.name AS teamname', 't.room',
-                'c.name AS catname',
-                's.cid', 'co.shortname',
-                'cp.shortname AS probshortname', 'cp.color',
-                'a.affilid AS affilid', 'a.shortname AS affilshort' )
-            ->from(Balloon::class, 'b')
-            ->leftJoin('b.submission', 's')
-            ->leftJoin('s.problem', 'p')
-            ->leftJoin('s.contest', 'co')
-            ->leftJoin('p.contest_problems', 'cp', Join::WITH, 'co.cid = cp.contest AND p.probid = cp.problem')
-            ->leftJoin('s.team', 't')
-            ->leftJoin('t.category', 'c')
-            ->leftJoin('t.affiliation', 'a')
-            ->andWhere('co.cid = :cid')
-            ->setParameter(':cid', $contest->getCid())
-            ->orderBy('b.done', 'ASC')
-            ->addOrderBy('s.submittime', 'DESC');
-
-        $balloons = $query->getQuery()->getResult();
-        // Loop once over the results to get totals and awards
-        $TOTAL_BALLOONS = $AWARD_BALLOONS = [];
-        foreach ($balloons as $balloonsData) {
-            if ( $balloonsData['color'] === null ) {
-                continue;
-            }
-
-            $TOTAL_BALLOONS[$balloonsData['teamid']][$balloonsData['probshortname']] = $balloonsData['color'];
-
-            // Keep a list of balloons that were first to solve this problem;
-            // can be multiple, one for each sortorder.
-            if (in_array($balloonsData['probid'], $firstSolvers[$balloonsData['teamid']]??[], true) ) {
-                $AWARD_BALLOONS['problem'][$balloonsData['probid']][] = $balloonsData[0]->getBalloonId();
-            }
-            // Keep overwriting this - in the end it'll
-            // contain the id of the first balloon in this contest.
-            $AWARD_BALLOONS['contest'] = $balloonsData[0]->getBalloonId();
-        }
-
-        // Loop again to construct table
-        $balloons_table = [];
-        foreach ($balloons as $balloonsData) {
-            $color = $balloonsData['color'];
+        $balloons_table = $balloonService->collectBalloonTable($contest);
 
-            if ( $color === null ) {
-                continue;
-            }
-            $balloon = $balloonsData[0];
-
-            $balloonId = $balloon->getBalloonId();
-
-            $stime = $balloonsData['submittime'];
-
-            if ( isset($freezetime) && $stime >= $freezetime) {
-                continue;
-            }
-
-            $balloondata = [];
-            $balloondata['balloonid'] = $balloonId;
-            $balloondata['time'] = $stime;
-            $balloondata['solved'] = Utils::balloonSym($color) . " " . $balloonsData['probshortname'];
-            $balloondata['color'] = $color;
-            $balloondata['problem'] = $balloonsData['probshortname'];
-            $balloondata['team'] = "t" . $balloonsData['teamid'] . ": " . $balloonsData['teamname'];
-            $balloondata['teamid'] = $balloonsData['teamid'];
-            $balloondata['location'] = $balloonsData['room'];
-            $balloondata['affiliation'] = $balloonsData['affilshort'];
-            $balloondata['affiliationid'] = $balloonsData['affilid'];
-            $balloondata['category'] = $balloonsData['catname'];
-
-            ksort($TOTAL_BALLOONS[$balloonsData['teamid']]);
-            $balloondata['total'] = $TOTAL_BALLOONS[$balloonsData['teamid']];
-
-            $comments = [];
-            if ($AWARD_BALLOONS['contest'] == $balloonId) {
-                $comments[] = 'first in contest';
-            } elseif (isset($AWARD_BALLOONS['problem'][$balloonsData['probid']])
-                   && in_array($balloonId, $AWARD_BALLOONS['problem'][$balloonsData['probid']], true)) {
-                $comments[] = 'first for problem';
-            }
-
-            $balloondata['awards'] = implode('; ', $comments);
-
-            if ( $balloon->getDone() ) {
+        // Add CSS class and actions.
+        foreach ($balloons_table as $element) {
+            if ($element['data']['done']) {
                 $cssclass = 'disabled';
                 $balloonactions = [[]];
-                $balloondata['done'] = true;
             } else {
                 $cssclass = null;
-                $balloondata['done'] = false;
                 $balloonactions = [[
                     'icon' => 'running',
                     'title' => 'mark balloon as done',
                     'link' => $this->generateUrl('jury_balloons_setdone', [
-                        'balloonId' => $balloon->getBalloonId(),
+                        'balloonId' => $element['data']['balloonid'],
                     ])]];
             }
-
-
-            $balloons_table[] = [
-                'data' => $balloondata,
-                'actions' => $balloonactions,
-                'cssclass' => $cssclass,
-            ];
+            $element['data']['actions'] = $balloonactions;
+            $element['data']['cssclass'] = $cssclass;
         }
 
         // Load preselected filters
@@ -217,7 +116,7 @@ class BalloonController extends AbstractController
                 'url' => $this->generateUrl('jury_balloons'),
                 'ajax' => true
             ],
-            'isfrozen' => $contestIsFrozen,
+            'isfrozen' => isset($contest->getState()['frozen']),
             'hasFilters' => !empty($filters),
             'filteredAffiliations' => $filteredAffiliations,
             'balloons' => $balloons_table
diff --git a/webapp/src/Service/BalloonService.php b/webapp/src/Service/BalloonService.php
index 512beaf4a..3e8bb9fd8 100644
--- a/webapp/src/Service/BalloonService.php
+++ b/webapp/src/Service/BalloonService.php
@@ -5,8 +5,11 @@ namespace App\Service;
 use App\Entity\Balloon;
 use App\Entity\Contest;
 use App\Entity\Judging;
+use App\Entity\ScoreCache;
 use App\Entity\Submission;
+use App\Utils\Utils;
 use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Query\Expr\Join;
 
 class BalloonService
 {
@@ -97,4 +100,111 @@ class BalloonService
             $this->em->flush();
         }
     }
+
+    public function collectBalloonTable(Contest $contest): array
+    {
+        $em = $this->em;
+        $showPostFreeze = (bool)$this->config->get('show_balloons_postfreeze');
+        if (!$showPostFreeze) {
+            $freezetime = $contest->getFreezeTime();
+        }
+
+        // Build a list of teams and the problems they solved first
+        $firstSolved = $em->getRepository(ScoreCache::class)->findBy(['is_first_to_solve' => 1]);
+        $firstSolvers = [];
+        foreach ($firstSolved as $scoreCache) {
+            $firstSolvers[$scoreCache->getTeam()->getTeamId()][] = $scoreCache->getProblem()->getProbid();
+        }
+
+        $query = $em->createQueryBuilder()
+            ->select('b', 's.submittime', 'p.probid',
+                't.teamid', 't.name AS teamname', 't.room',
+                'c.name AS catname',
+                's.cid', 'co.shortname',
+                'cp.shortname AS probshortname', 'cp.color',
+                'a.affilid AS affilid', 'a.shortname AS affilshort')
+            ->from(Balloon::class, 'b')
+            ->leftJoin('b.submission', 's')
+            ->leftJoin('s.problem', 'p')
+            ->leftJoin('s.contest', 'co')
+            ->leftJoin('p.contest_problems', 'cp', Join::WITH, 'co.cid = cp.contest AND p.probid = cp.problem')
+            ->leftJoin('s.team', 't')
+            ->leftJoin('t.category', 'c')
+            ->leftJoin('t.affiliation', 'a')
+            ->andWhere('co.cid = :cid')
+            ->setParameter(':cid', $contest->getCid())
+            ->orderBy('b.done', 'ASC')
+            ->addOrderBy('s.submittime', 'DESC');
+
+        $balloons = $query->getQuery()->getResult();
+        // Loop once over the results to get totals and awards
+        $TOTAL_BALLOONS = $AWARD_BALLOONS = [];
+        foreach ($balloons as $balloonsData) {
+            if ($balloonsData['color'] === null) {
+                continue;
+            }
+
+            $TOTAL_BALLOONS[$balloonsData['teamid']][$balloonsData['probshortname']] = $balloonsData['color'];
+
+            // Keep a list of balloons that were first to solve this problem;
+            // can be multiple, one for each sortorder.
+            if (in_array($balloonsData['probid'], $firstSolvers[$balloonsData['teamid']] ?? [], true)) {
+                $AWARD_BALLOONS['problem'][$balloonsData['probid']][] = $balloonsData[0]->getBalloonId();
+            }
+            // Keep overwriting this - in the end it'll
+            // contain the id of the first balloon in this contest.
+            $AWARD_BALLOONS['contest'] = $balloonsData[0]->getBalloonId();
+        }
+
+        // Loop again to construct table
+        $balloons_table = [];
+        foreach ($balloons as $balloonsData) {
+            $color = $balloonsData['color'];
+
+            if ($color === null) {
+                continue;
+            }
+            $balloon = $balloonsData[0];
+
+            $balloonId = $balloon->getBalloonId();
+
+            $stime = $balloonsData['submittime'];
+
+            if (isset($freezetime) && $stime >= $freezetime) {
+                continue;
+            }
+
+            $balloondata = [];
+            $balloondata['balloonid'] = $balloonId;
+            $balloondata['time'] = $stime;
+            $balloondata['solved'] = Utils::balloonSym($color) . " " . $balloonsData['probshortname'];
+            $balloondata['color'] = $color;
+            $balloondata['problem'] = $balloonsData['probshortname'];
+            $balloondata['team'] = "t" . $balloonsData['teamid'] . ": " . $balloonsData['teamname'];
+            $balloondata['teamid'] = $balloonsData['teamid'];
+            $balloondata['location'] = $balloonsData['room'];
+            $balloondata['affiliation'] = $balloonsData['affilshort'];
+            $balloondata['affiliationid'] = $balloonsData['affilid'];
+            $balloondata['category'] = $balloonsData['catname'];
+
+            ksort($TOTAL_BALLOONS[$balloonsData['teamid']]);
+            $balloondata['total'] = $TOTAL_BALLOONS[$balloonsData['teamid']];
+
+            $comments = [];
+            if ($AWARD_BALLOONS['contest'] == $balloonId) {
+                $comments[] = 'first in contest';
+            } elseif (isset($AWARD_BALLOONS['problem'][$balloonsData['probid']])
+                && in_array($balloonId, $AWARD_BALLOONS['problem'][$balloonsData['probid']], true)) {
+                $comments[] = 'first for problem';
+            }
+
+            $balloondata['awards'] = implode('; ', $comments);
+            $balloondata['done'] = $balloon->getDone();
+
+            $balloons_table[] = [
+                'data' => $balloondata,
+            ];
+        }
+        return $balloons_table;
+    }
 }
-- 
2.20.1

